core/modules/book/book.module

```diff
@@ -662,6 +662,11 @@ function _book_update_outline(Node $node) {
         backdrop_static_reset('book_get_books');
       }
     }
+    // Get all nodes that are part of this book and flush their caches.
+    $affected_nids = db_query("SELECT nid FROM {book} WHERE bid = :bid", array(
+      ':bid' => $node->book['bid'],
+    ))->fetchAllAssoc('nid');
+    entity_get_controller('node')->resetCache(array_keys($affected_nids));
```

Comment:
This looks OK, but my concern about this is that `_book_update_outline()` is called multiple times when you delete a book page, each time for each of the children (which may mean a LOT of times if you delete the top-most node). Therefore we'd be running this query for each of the children, and reseting the cache for the same nids over and over. While the approach here reuses code, it seems wasteful in that situation.

An alternative may be to create a new function (e.g. `_book_clear_cache_entity_node($node)`) that gets called from `book_node_insert()`, `book_node_update()` and `book_node_predelete()` only once.

Reply:
> ... is called multiple times when you delete a book page

Ah, that's a valid concern. Have to take a closer look re node deletion - in which case this call is useless. :+1:

Reply:
My idea was, that if a book gets deleted, we wouldn't end up here, as `empty($node->book['bid'])` should catch that. And then the outline isn't rebuilt. :thinking:

Reply:
Fancy Book module. :wink: 

@argiepiano The loop in book_node_predelete() was probably your concern, but that behaves correctly. Function _book_update_outline() runs once for every newly created "stub" book, if a top level node gets deleted.

Example:
```
- The book
- - Subpage 1
- - - Below subpage 1
- - Subpage 2
- - - Below subpage 2
```
When deleting "The book", two structures get created, one from "Subpage 1" with its children, one from "Subpage 2" with its children. So inside the loop the book ID is a different one and child nodes are different, too. No overlap as far as I can see. Or am I missing something?

Reply:
Right! This looks good then!